compilers/concrete-compiler/compiler/lib/Dialect/FHELinalg/Transforms/Tiling.cpp

```diff
@@ -20,315 +25,148 @@
-
-    mlir::arith::ConstantIndexOp V =
-        rewriter.create<mlir::arith::ConstantIndexOp>(origLoc, iV);
-
-    // Lower bound for all for loops
-    mlir::arith::ConstantIndexOp lb =
-        rewriter.create<mlir::arith::ConstantIndexOp>(origLoc, 0);
-
-    // Upper bounds are determined by the size of the operands
-    mlir::arith::ConstantIndexOp ubT =
-        rewriter.create<mlir::arith::ConstantIndexOp>(origLoc,
-                                                      ATTy.getDimSize(0));
-
-    mlir::arith::ConstantIndexOp ubU =
-        rewriter.create<mlir::arith::ConstantIndexOp>(origLoc,
-                                                      ATTy.getDimSize(1));
-
-    mlir::arith::ConstantIndexOp ubV =
-        rewriter.create<mlir::arith::ConstantIndexOp>(origLoc,
-                                                      BTTy.getDimSize(1));
-
-    // Bounds and steps in vector form
-    llvm::SmallVector<mlir::Value, 3> lbs{lb, lb, lb};
-    llvm::SmallVector<mlir::Value, 3> ubs{ubT, ubU, ubV};
-    llvm::SmallVector<mlir::Value, 3> steps{T, U, V};
-
-    // Callback function to build the body of the innermost loop
-    auto innermostBodyBuilder = [&](mlir::OpBuilder &builder,
-                                    mlir::Location location,
-                                    mlir::ValueRange inductionVars,
-                                    mlir::ValueRange iterArgs) {
-      // TxU tile from A
-      mlir::tensor::ExtractSliceOp ATile = extractContiguous2DSlice(
-          builder, origLoc, A, {iT, iU}, {inductionVars[0], inductionVars[1]});
-      // UxV tile from B
-      mlir::tensor::ExtractSliceOp BTile = extractContiguous2DSlice(
-          builder, origLoc, B, {iU, iV}, {inductionVars[1], inductionVars[2]});
-
-      // TxV tile from C
-      mlir::tensor::ExtractSliceOp CTile = extractContiguous2DSlice(
-          builder, origLoc, *iterArgs.begin(), {iT, iV},
-          {inductionVars[0], inductionVars[2]});
-
-      // Multiplication of the tiles
-      mlir::concretelang::FHELinalg::MatMulEintIntOp tiledMul =
-          builder.create<mlir::concretelang::FHELinalg::MatMulEintIntOp>(
-              origLoc,
-              mlir::RankedTensorType::get(llvm::SmallVector<int64_t, 2>{iT, iV},
-                                          CTTy.getElementType()),
-              ATile, BTile);
-
-      // Mark matrix multiplication to prevent recursive
-      // application of the rewriting pattern
-      tiledMul.getOperation()->setAttr(kTransformMarker,
+    // If all, but the last iterator types are parallel and the last
+    // type is a reduction, tile the reduction
+    if (iteratorTypes.size() > 1 &&
+        std::all_of(iteratorTypes.begin(), iteratorTypes.end() - 1,
+                    [](mlir::utils::IteratorType itty) {
+                      return itty == mlir::utils::IteratorType::parallel;
+                    }) &&
+        *(iteratorTypes.end() - 1) == mlir::utils::IteratorType::reduction) {
```

Comment:
Not sure if I got the comment correctly. If you are referring to the fact that the cast to `mlir::PartialReductionOpInterface` is not checked: this cannot fail, since `op` is a `linalg.generic` op, which implements this interface.